File: C++/Heap/MergeKSortedArrays.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct Triplet
{
    int ele, arrPos, elePos;
    Triplet(int element, int arrPosition, int elePosition)
    {
        ele = element;
        arrPos = arrPosition;
        elePos = elePosition;
    }
};

struct Comparator
{
    bool operator()(Triplet st1, Triplet st2)
    {
        return st1.ele > st2.ele;
    }
};

vector<int> mergeArrays(vector<vector<int>> &arr)
{
    vector<int> res;
    priority_queue<Triplet, vector<Triplet>, Comparator> pq;
    for (int i = 0; i < arr.size(); i++)
    {
        Triplet t(arr[i][0], i, 0);
        pq.push(t);
    }
    while (!pq.empty())
    {
        Triplet curr = pq.top();
        pq.pop();
        res.push_back(curr.ele);
        int arrPos = curr.arrPos, elePos = curr.elePos;
        if (elePos + 1 < arr[arrPos].size())
        {
            Triplet t(arr[arrPos][elePos + 1], arrPos, elePos + 1);
            pq.push(t);
        }
    }
    return res;
}

// int main()
// {
//     vector<vector<int>> arr = {
//         {5, 10},
//         {4, 9},
//         {5, 6},
//     };
//     for (auto i : mergeArrays(arr))
//         cout << i << " ";
//     cout << endl;
// }

File: Python/heap/merge_k_sorted_arrays.py
from queue import PriorityQueue


class Triplet:
    def __init__(self, element, arr_position, ele_position):
        self.element = element
        self.arr_position = arr_position
        self.ele_position = ele_position

    def __lt__(self, other):
        return self.element < other.element

    # def __eq__(self, other):
    #     return self.element == other.element

    # def __le__(self, other):
    #     return self.element <= other.element

    # def __ge__(self, other):
    #     return self.element >= other.element

    # def __ne__(self, other):
    #     return self.element != other.element


def merge_k_sorted_arrays(arr):
    res = []
    pq = PriorityQueue()
    for i in range(len(arr)):
        pq.put(Triplet(arr[i][0], i, 0))
    while not pq.empty():
        curr = pq.get()
        res.append(curr.element)
        arr_pos, ele_pos = curr.arr_position, curr.ele_position
        if ele_pos + 1 < len(arr[arr_pos]):
            pq.put(Triplet(arr[arr_pos][ele_pos + 1], arr_pos, ele_pos + 1))
    return res


if __name__ == "__main__":
    arr = [[5, 10], [4, 9], [5, 6]]
    print(merge_k_sorted_arrays(arr))
